distsys/resources/tcpmailboxes.go

```diff
@@ -159,17 +199,24 @@ func (res *tcpMailboxesLocalArchetypeResource) handleConn(conn net.Conn) {
 			localBuffer = nil
 			hasBegun = true
 		case tcpNetworkValue:
+			if res.closing.get() {
+				continue
+			}
 			var value tla.TLAValue
 			err = decoder.Decode(&value)
 			if err != nil {
 				continue
 			}
 			localBuffer = append(localBuffer, value)
 		case tcpNetworkPreCommit:
+			if res.closing.get() {
+				continue
+			}
 			err = encoder.Encode(struct{}{})
 			if err != nil {
 				continue
 			}
```

Comment:
what if `wg.Wait()` gets called by Close during this gap, assuming the send completed already?

It seems something like this is possible (assuming one active connection, and a concurrent call to Close):
- the connection reads `closing`, gets `false`, and responds to a PreCommit
- Close sets `closing` to `true`
- Close calls `Wait()` on the waitgroup, whose value is `0` so `Wait()` returns immediately
- the connection runs `Add(1)` on the waitgroup
- Close starts tearing things down
- :boom: ?

Reply:
I think it's necessary to track the Commit critical section more pre-emptively, so this gap doesn't exist...

maybe opening the waitgroup before even checking `closing`, and closing it again if the `closing` check yields `true`?

Reply:
Concurrent calls to `Close` is not possible since `mpcalCtx.Close` calls it only once (should document this). However, the scenario you have mentioned is possible. I'll fix it.

Reply:
(maybe a wording point, but when I said "concurrent" I meant relative to the active connection, not that one would call Close more than once)

sounds good though

Reply:
I see, my bad understanding.

Reply:
follow-up on the new changes:

I think it probably fixes the issue, but, before thinking so deeply, I notice that this change basically single-threads the entire mailbox server: all network ops will be stuck in the same critical section, even the ones that are safe to run concurrently (most of them).

It seems we're using a lot more locking than is actually needed here - the only concurrency issue is that shutdown might interfere with an ongoing connection running a commit.
A simple sequencing point might be (still using an atomic `closing` variable):
- Close sets `closing = true`
- the `listen()` goroutine guarantees that once `closing` is set, no more connections will be accepted (so, even if a new connection is made at this point, nothing happens)
- whenever a new connection is made, while holding the `closing` lock, the wait group is `Add(1)`
- now, the `Wait()` in Close will block until _all connections are closed_
- each connection can poll `closing`, and, unless it's processing a Commit message, shut down immediately, decrementing the wait group
- since "no connections running" >= "no commits running", we are safe
- all the connections can still be processed async, because they will be polling, rather than interrupted, so no internal synchronisation is needed beyond atomicity for `closing`

thinking more, this does add the issue of all connection ops being blocking... liveness could be retained by using timeouts, like the remote mailboxes?

it might be simpler to do this via call, given how much I've written already

Reply:
(explained via call; I had missed RWLock, and a few other issues)